### PR TITLE
Adds germplasm collection panel

### PR DESCRIPTION
**Pull request by reynoldtan**
This PR will add germplasm collection panel (list exportable to a csv file) to study/experiment content.

## Metadata

 - Issue #

- [x] I feel this PR is ready to be merged.
- [ ] This PR is dependent upon [PR #/ nothing]

Documentation:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Description

1. Modified local__project_germcollection field to
a. Fetch germplasm/stock collection of a study
b. Create table markup showing germplasm list based on the set collection in the field settings panel.
c. Update install file to implement a hook schema and setup backend used to map collection to a study.
 
## Testing?

- [ ] I tested on a generic Tripal Site
- [x] I tested on a KnowPulse Clone
- [ ] This PR includes automated testing

1. Clone module and switch branch to project-collection-field.
2. Install and enable module - germcollection under Tripal Germplasm
3. Modify Study Tripal Content Type and add fields germplasm collection (local: project germcollection)
Screen Shot 2022-03-02 at 12 16 28 PM
Click check for new field link if fields have not been added previously.
4. Click save. Note: before saving make sure fields are not in the disabled list in managed display page of the field and set the FORMAT to either inline or above and not hidden (shown below).
Screen Shot 2022-03-03 at 10 02 06 AM

6. Modify an existing study and set the value under Germplasm field in the update study page.
Screen Shot 2022-03-02 at 12 19 41 PM
7. Click save and reload study page to view germplasm collection list.
Screen Shot 2022-03-02 at 12 20 57 PM

**Additional configuration:**
1. To view and export germplasm collection, user must be logged in and given permission to view and download germplasm collection.

### CODE REVIEW

**Comment by ruobinLiu** [comment]
I have the add germplasm collection panel tested on multiple projects, it works as expected overall. Except sometimes, the germplasm only shows up when a page is refreshed through  "put cursor beside the url link", then "click enter button".

File: germcollection/api/germcollection.api.inc [hunk 1]
```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * @file
+ * Contains general API required by this field.
+ * 
+ * @param $collection_id
+ *   Integer, collection id number.
+ * 
+ * @return array
+ *   Germplasm collection where each item in the array is - Name, Accession, Genus, Species and Origin (Country of origin).
```
Comment (laceysanderson) [inline germcollection/api/germcollection.api.inc]:
```suggestion
 *   Germplasm collection where each item in the array is - Name, Accession, Entity ID, Genus, Species and Origin (Country of origin).
```

The entity_id is added to the results on ln49 so adding it to the documentation.

File: germcollection/api/germcollection.api.inc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ *   Integer, type id of stock corresponding to a term in chado.cvterm.
+ *
+ * @return
+ *   Entity id or # or an blank anchor when entity id cannot be extracted.
+ */
+function germcollection_get_entity_id($stock_id, $type_id) {
+  // This character is also used to tell JS if node should be rendered as link or a text.
+  $entity_id = '#';
+
+  $m = db_query('SELECT t2.name
```
Comment (laceysanderson) [inline germcollection/api/germcollection.api.inc:76]:
```suggestion
  // Retrieves the name of the Tripal bundle based on the stock type_id
  $m = db_query('SELECT t2.name
```

File: germcollection/api/germcollection.api.inc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    WHERE t1.accession = (
+      SELECT t3.accession
+      FROM chado.dbxref AS t3 INNER JOIN chado.cvterm AS t4 USING(dbxref_id)
+      WHERE t4.cvterm_id = :stock_type)', array(':stock_type' => $type_id))
+    ->fetchField();
+
+  if ($m) {
+    $chado_bundle = 'chado_' . $m;
+
+    $sql = sprintf('SELECT t1.entity_id FROM %s AS t1 INNER JOIN {stock} AS t2 ON t1.record_id = t2.stock_id
```
Comment (laceysanderson) [inline germcollection/api/germcollection.api.inc]: This is an unnecessarily long/slow query. You can skip the inner join here and just say `t1.record_id = :stock_id`.

**Review by laceysanderson — CHANGES_REQUESTED** [review]
See further comments on the code review -Overall good though :-)

This field should also include the name of the germplasm collection and a link to a page (although we're still working on those so ensure it only shows if the page exists as always :-) ). The idea is that once this field is in place the old text field addition of stock collection would go away.

With this change, we will always want the name + number of germplasm shown regardless of permission. If no permission though the download link would be disabled and no list would be shown.

**Review by laceysanderson — CHANGES_REQUESTED** [review]
Items from my original review:
- [x] comment additions
- [x] remove inner join on query for performance reasons
- [ ] add germplasm collection name + link to page in field
- [ ] always show name + number of germplasm regardless of permission. If no permission then the list and download link are not shown.

So it looks like the last two items were not done yet?

> add germplasm collection name + link to page in field

We want the name of the germplasm collection with a link to its page provided in This Field -specifically through the formatted. Currently you see the name in the page due to the following field which just provides a textfield for a projectprop entry. This is clearly not ideal and we want to remove the chado property field completely.
Screen Shot 2022-04-13 at 4 48 32 PM

To do that though your field needs to have the collection name with a link to the collection page. My original field did have this but you removed it... It's still stored in the `$link` variable that you no longer use.

> always show name + number of germplasm regardless of permission. If no permission then the list and download link are not shown.

Right now line 91 of the formatted ensures the field shows "Germplasm collection not available" if people do not have permission to see germplasm. However, it should still show the information in `$field_markup` but with $download_link and $summary_tables removed. **The germplasm collection name/link and germplasm count should ALWAYS SHOW regardless of permission.**

**Comment by reynoldtan** [comment]
Revised field as describe. Tested:

- [x] user given permission ✔️ 
- [x] user has no permission ✖️ 
- [x] anonymous user ✖️ 

image

**Review by laceysanderson — APPROVED** [review]
Changes look good! I am merging thus but we shouldn't make it live until July just to ensure nothing goes wrong right before the workshop ;-p

**Pull request merged**